### PR TITLE
fix(config): fixing Compression parsing with tests

### PR DESCRIPTION
**Pull request by fanatid**
Moved from https://github.com/timberio/vector/pull/5281#discussion_r536383664

### CODE REVIEW

File: src/sinks/util/buffer/compression.rs [hunk 1]
```diff
@@ -99,9 +100,34 @@ impl<'de> de::Deserialize<'de> for Compression {
...
                             }
-                            level = Some(match map.next_value::<usize>() {
-                                Ok(value) => value.to_string(),
-                                Err(_) => map.next_value::<&str>()?.to_string(),
+                            level = Some(match map.next_value::<Value>()? {
+                                Value::Number(level) => match level.as_u64() {
+                                    Some(value) if value <= 9 => value as usize,
+                                    Some(_) | None => {
+                                        return Err(de::Error::invalid_value(
+                                            de::Unexpected::Other("0, 1, 2, 3, 4, 5, 6, 7, 8 or 9"),
```
Comment (juchiast) [inline src/sinks/util/buffer/compression.rs:108]: First parameter of `invalid_value` is the invalid value in the input, so we need to pass `level` here. I'm not sure what to do with the `None` case, maybe we convert it to a string if it is too cumbersome to conver `serde_json::Number` into `Unexpected`. 
 
Second parameter is what data was expected, using `self` as the second parameter will print "expected string or map". You can use a reference to a string here, so `&"0, 1, 2, 3, 4, 5, 6, 7, 8 or 9"` would be perfect.

File: src/sinks/util/buffer/compression.rs [hunk 2]
```diff
@@ -99,9 +100,34 @@ impl<'de> de::Deserialize<'de> for Compression {
...
+                                        ))
+                                    }
+                                },
+                                Value::String(level) => match level.as_str() {
+                                    "none" => GZIP_NONE,
+                                    "fast" => GZIP_FAST,
+                                    "default" => GZIP_DEFAULT,
+                                    "best" => GZIP_BEST,
+                                    _ => {
+                                        return Err(de::Error::invalid_value(
```
Comment (juchiast) [inline src/sinks/util/buffer/compression.rs:119]: Same problem here.

File: src/sinks/util/buffer/compression.rs [hunk 3]
```diff
@@ -99,9 +100,34 @@ impl<'de> de::Deserialize<'de> for Compression {
...
+                                            de::Unexpected::Other("none, fast, default or best"),
+                                            &self,
+                                        ))
+                                    }
+                                },
+                                _ => {
+                                    return Err(de::Error::invalid_type(
+                                        de::Unexpected::Other("Number or String"),
```
Comment (juchiast) [inline src/sinks/util/buffer/compression.rs:127]: Same problem here. I'm not sure how to get an appropriate `Unexpected` variant here.

**Comment by juchiast** [comment]
Not introduced in this PR, but error message on `visit_str` is confusing:
```
invalid value: string "asdf", expected string or map at line 1 column 6
```

You can use
```rust
                    _ => Err(de::Error::invalid_value(
                        de::Unexpected::Str(s),
                        &r#""none" or "gzip""#,
                    )),
```

with will print
```
invalid value: string "asdf", expected none or gzip
````

Using `unknown_variant` is also fine, it will print 
```
unknown variant `asdf`, expected `none` or `gzip`
```

**Review by juchiast — CHANGES_REQUESTED** [review]

**Comment by fanatid** [comment]
@juchiast thanks! I changed errors and added tests for them.

**Review by juchiast — APPROVED** [review]
Look good! Some `format!("{}", ...)` could be replaced with `to_string` to avoid macro invocation.

**Comment by fanatid** [comment]
@juchiast thank you for review!

**Pull request merged**